### ISSUE TITLE
types can't be assigned to supersets

### ISSUE THREAD

**Issue by cdmistman** [microsoft/pyright#7603]
```python
from typing import Optional, TypedDict


class Foo(TypedDict):
  field: Optional[str]

class Bar(TypedDict):
  field: str

x: Bar = { 'field' : 'hello' }
y: Foo = x
```

i get the error:

```
Expression of type "Bar" cannot be assigned to declared type "Foo"
  "field" is an incompatible type
    Type "str" cannot be assigned to type "str | None"
      "str" is incompatible with "None"
```

**Comment by erictraut** [comment]
This behavior is correct and in compliance with the [typing spec](https://typing.readthedocs.io/en/latest/spec/typeddict.html) and [PEP 589](https://peps.python.org/pep-0589/), which introduced `TypedDict` to the Python type system.

The type `Bar` is not type compatible with type `Foo` because `Bar` allows `field` to contain `None` whereas `Foo` does not.

Consider if you were to add the line `y["field"] = None` to the end of your code sample. This statement would be allowed because the `field` item in `Foo` allows the value to be `None`, but it would violate the type of `x` because `Bar` does not allow `field` to have a value of `None`.

**Comment by erictraut** [comment]
Note that if you construct a new dict object, it will type check without errors.

```python
x: Bar = {"field": "hello"}
y: Foo = {**x}

# This no longer presents a problem because `y`
# is a separate object from `x`.
y["field"] = None
```
